src/communication/rpc_impl.rs

```diff
@@ -0,0 +1,486 @@
+            )))
+        }
+    }
+
+    fn handle_response(&self, reqid: &UUID, response_message: UMessage) {
+        if let Ok(mut pending_requests) = self.pending_requests.lock() {
+            if let Some(sender) = pending_requests.remove(reqid) {
+                let _ = sender.send(response_message);
+            } else {
+                // we seem to have received a duplicate of the response message,
+                // ignoring it
+            }
+        }
+    }
+
+    fn remove_pending_request(&self, reqid: &UUID) -> Option<Sender<UMessage>> {
+        self.pending_requests
+            .lock()
+            .map_or(None, |mut pending_requests| pending_requests.remove(reqid))
+    }
+
+    #[cfg(test)]
+    fn contains(&self, reqid: &UUID) -> bool {
+        self.pending_requests
+            .lock()
+            .map_or(false, |pending_requests| {
+                pending_requests.contains_key(reqid)
+            })
+    }
+}
+
+#[async_trait]
+impl UListener for ResponseListener {
+    async fn on_receive(&self, msg: UMessage) {
+        if msg.is_response() {
+            if let Some(reqid) = msg
+                .attributes
+                .as_ref()
+                .and_then(|attribs| attribs.reqid.clone().into_option())
+            {
+                self.handle_response(&reqid, msg);
+            }
+        }
+    }
+}
+
+/// An ['RpcClient'] which keeps all information about pending requests in memory.
+pub struct InMemoryRpcClient {
+    transport: Arc<dyn UTransport>,
+    uri_provider: Arc<dyn LocalUriProvider>,
+    response_listener: Arc<ResponseListener>,
+}
+
+impl InMemoryRpcClient {
+    /// Creates a new RPC client for a given transport.
+    ///
+    /// # Arguments
+    ///
+    /// * `transport` - The uProtocol Transport Layer implementation to use for invoking service operations.
+    /// * `uri_provider` - The helper for creating URIs that represent local resources.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the generic RPC Response listener could not be
+    /// registered with the given transport.
+    pub async fn new(
+        transport: Arc<dyn UTransport>,
+        uri_provider: Arc<dyn LocalUriProvider>,
+    ) -> Result<Self, UStatus> {
+        let response_listener = Arc::new(ResponseListener {
+            pending_requests: Mutex::new(HashMap::new()),
+        });
+        transport
+            .register_listener(
+                &UUri::any(),
+                Some(&uri_provider.get_source_uri()),
+                response_listener.clone(),
+            )
+            .await?;
+
+        Ok(InMemoryRpcClient {
+            transport,
+            uri_provider,
+            response_listener,
+        })
+    }
+
+    #[cfg(test)]
+    fn contains_pending_request(&self, reqid: &UUID) -> bool {
+        self.response_listener.contains(reqid)
+    }
+}
+
+#[async_trait]
+impl RpcClient for InMemoryRpcClient {
+    async fn invoke_method(
+        &self,
+        method: UUri,
```

Comment:
I'm not buying or selling here.

Seems like the [up-cpp](https://github.com/eclipse-uprotocol/up-cpp/blob/3b80662d884b189d36114ef2922a92bc3699a397/include/up-cpp/communication/RpcClient.h#L51) version bakes their method into the construction of their RpcClient vs here we do not and thus are able to use a single RpcClient for multiple different service method invocation types.

Wondering if you had some thoughts on the pros/cons.

Tagging @gregmedd for any thoughts as well.

Reply:
Not really but I guess when baking it into the constructor, there is much less dynamic memory allocation required during runtime, which might be desirable? Assuming that a client uEntity will have a fixed set of other uEntities it needs to interact with, having dedicated RpcClient instances for them should also be ok.

Reply:
Yeah, frankly I think there's no one true right way here.

In the case of having it in the constructor, I think this fits the use-case pretty nicely of 99% of users of `RpcClient` and as you said _may_ reduce dynamic memory allocation.

On the other hand, having the _option_ for a single `RpcClient` is not a bad thing.

I suppose we could have some examples in here or in a future PR showing both styles to inspire users to the direction that makes sense to them.

Looking to see if @gregmedd has any thoughts here given that I now know that up-cpp is the outlier :^)

Reply:
The two approaches are different philosophies. In `up-cpp`, we have applied the philosophy that one instance of a layer 2 object represents one association - this has been applied evenly across all of our layer 2 APIs. Specifically speaking to `RpcClient`, we expect the usage pattern to look something like:

```C++
class SomeRemoteEntity {
    ...
    RpcClient resource_foo;
    RpcClient resource_bar;
    ...
public:
    SomeRemoteEntity(/* options */) : resource_foo(/* setup foo */), resource_bar(/*setup bar*/) {}

    bool foo(/* some list of parameters */) {
        PayloadBuilder parameters;
        // Prepare parameters as payload
        ...
        auto future = resource_foo.invokeMethod(parameters);
        // Evaluate return value of future
        ...
        return foo_result;
    }
    ...
};
```

The setup for the RPC method occurs once, gets wrapped by some appropriate higher layer, then is reused for the life of the client entity. With that usage pattern, it make sense to have one `RpcClient` instance represent the association with one remote `RpcServer`'s method. In the example above, most code within a uE would call the `foo()` method on `SomeRemoteEntity` instead of using `RpcClient::invokeMethod()` directly since additional work is needed to prepare a request and process a response. This also has some other advantages, such as being able to track which pending requests belong to which `RpcClient`, allowing us to validate parameters earlier (i.e. during startup) rather than later (i.e. at the moment the method is invoked), and making the point of ownership for the instance more clear.

The alternative is to instead only require (at least) one `RpcClient` instance for a uE - a singleton-like `RpcClient` - where each call to `RpcClient::invokeMethod()` can invoke any possible remote method. To wrap those methods like the example above, the parameters needed to invoke the remote RPC method would need to be provided at each point of use. This makes it easier to reuse resources across the uses of `RpcClient` (such as a callback registration), with one of the tradeoffs being that validation is deferred to invocation.

In the `up-cpp` implementation of `RpcClient`, I plan to handle the sharing of some resources inside the `RpcClient` itself. The author of the uE doesn't need to worry about that detail - they can just create and hold `RpcClient` instances for RPC methods they want to invoke.

Reply:
> In the example above, most code within a uE would call the `foo()` method on `SomeRemoteEntity` instead of using `RpcClient::invokeMethod()` directly since additional work is needed to prepare a request and process a response.

This also seems to be more amenable to some of the code generation tools that @stevenhartley has discussed, perhaps? WDYT Steven?

In addition, I really like this:

> allowing us to validate parameters earlier (i.e. during startup) rather than later (i.e. at the moment the method is invoked)

Reply:
On the other hand it also means that you have to use the same parameters for each invocation, e.g. the same TTL, priority, token etc. Not sure if this flexibility is needed/helpful, though.

I am fine with both approaches.
@PLeVasseur @AnotherDaniel do you have any preference?

Reply:
I like the idea of supporting the 99% use case a little more obviously and front-loading validation upon construction, so I'm leaning towards the approach of having the service method, ttl, etc supplied upon construction.

Anyone else have thoughts? @evshary?

Reply:
I actually prefer the more flexible method, where the parameters are not front-loaded. As per current experience in the usubscription code base that I'm working on - there are some places where this would be a major nuisance, as I need dynamic parameters (e.g. target UUris) every time.

So if we do a front-load version, I strongly suggest that this becomes a wrapper around the more generic one, and both are available for use (which probably would result in another source of confusion though...).

Reply:
I don't have strong opinions on this. It depends on how users use it. If users need to send requests to several methods, the dynamic one makes more sense. On the other hand, if users only stick to the same RPC Server, it's better to keep the front-loaded parameters. I might agree with @AnotherDaniel, since he is the user now :laughing:

Reply:
The let's merge this for now and the other variant later, if we have a clear picture of when it would be more advantageous.